core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -0,0 +1,755 @@
+ * @since 11.0
+ * @param <K> key type for the store
+ * @param <V> value type for the store
+ */
+public class AsyncNonBlockingStore<K, V> extends DelegatingNonBlockingStore<K, V> implements Consumer<Flowable<AsyncNonBlockingStore.Modification>> {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private static final boolean trace = log.isTraceEnabled();
+   private final NonBlockingStore<K, V> actual;
+
+   // Any non-null value can be passed to `onNext` when a new batch should be submitted - A value should only be
+   // submitted to this if the `batchFuture` was null and the caller was able to assign it to a new value
+   private final MulticastProcessor<Object> requestFlowable;
+   // Submit new Modifications to this on every write
+   private volatile FlowableProcessor<Modification> submissionFlowable;
+   // Closing this will stop the subsmissionFlowable subscription
+   private volatile Disposable startSub;
+
+   private Executor nonBlockingExecutor;
+   private int segmentCount;
+   private int modificationQueueSize;
+   private PersistenceConfiguration persistenceConfiguration;
+   private AsyncStoreConfiguration asyncConfiguration;
+
+   // "Non blocking" scheduler used for the purpose of delaying retry batch operations on failures
+   private ScheduledExecutorService scheduler;
+
+   // This variable will be non null if there is a pending batch being sent to the underlying store
+   // If a request causes the modification queue to overflow it will receive a stage back that is only complete
+   // when this future is completed (aka. previous replication has completed)
+   @GuardedBy("this")
+   private CompletableFuture<Void> batchFuture;
+
+   // This variable will be non null if the underlying store has been found to be not available
+   // Note that the async store will still be available as long as the queue size (ie. modificationMap.size) is not
+   // greater than the configured modificationQueueSize
+   @GuardedBy("this")
+   private CompletableFuture<Void> delegateAvailableFuture;
+
+   // Any pending modifications will be enqueued in this map
+   @GuardedBy("this")
+   private Map<Object, Modification> pendingModifications = new HashMap<>();
+   // If there is a pending clear this will be true
+   @GuardedBy("this")
+   private boolean hasPendingClear;
+   // The next two variables are held temporarily until a replication of the values is complete. We need to retain
+   // these values until we are sure the entries are actually in the store - note these variables are only written to
+   // via reference (thus the map is safe to read outside of this lock, but the reference must be read in synchronized)
+   // This map contains all the modifcations currently being replicated to the delegating store
+   @GuardedBy("this")
+   private Map<Object, Modification> replicatingModications;
+   // True if their is an oustanding clear that is being ran on the delegating store
```

Comment:
typos? also, does it make sense to change  `is being ran` => `is running`? 
```suggestion
   // True if there is an outstanding clear that is being ran on the delegating store
```